### ISSUE TITLE
Large number of `queryRows` requests when emphasizing or opening another viewport

### ISSUE THREAD

**Issue by a-gagnon** [iTwin/presentation#793]
Hi,
    since we upgraded to tree-widget-react ^3.0, we've noticed some performance hits when using our civil tools. After digging a little, we found that when elements are emphasized as part of our tool workflow, there is a large number of `IModelReadRpcInterface-3.7.0-queryRows` requests being performed. It also happens on an even larger scale when we open our section viewport.

It ultimately stabilizes as we keep playing with the tools and the views (most likely due to caching), but we have had a few complaints of performance degradation already, and even in localhost there is a noticeable performance hit from whatever was done in tree-widget-react: 2.x.

I have attached a video that shows the number of requests after a cold start of the page.
The first part when I open the `2-point cross-section` tool is to validate that if no emphasizing is involved, there are no `queryRows` requests being made.
At 0:30, the `Perpendicular cross section` tool is started, which will emphasize all alignments in the model.
At 0:45, we open the secondary viewport which triggers its own list of requests.

Please reach out if you need more info.
Thanks,
Alex

Attached video:
https://github.com/user-attachments/assets/518c24fa-c55f-4501-8678-9c975190fdda

**Comment by a-gagnon** [comment]
@grigasp 

**For 1), It's a bunch of** 
```
WITH RECURSIVE CategoryElements(id) AS (SELECT ECInstanceId id FROM BisCore.GeometricElement3d WHERE Model.Id = ? AND Category.Id = ? AND Parent IS NULL UNION ALL SELECT c.ECInstanceId id FROM BisCore.GeometricElement3d c JOIN CategoryElements p ON c.Parent.Id = p.id) SELECT COUNT(*) FROM CategoryElements
```

**For 2), almost all of them are the same as 1), but I did notice a very limited amount of other queries:**
```
SELECT s.ECInstanceId id, s.Parent.Id parentId, (SELECT m.ECInstanceId FROM bis.GeometricModel3d m WHERE m.ECInstanceId = HexToId(json_extract(s.JsonProperties, '$.Subject.Model.TargetPartition')) AND NOT m.IsPrivate AND EXISTS (SELECT 1 FROM BisCore.GeometricElement3d WHERE Model.Id = m.ECInstanceId)) targetPartitionId, CASE WHEN (json_extract(s.JsonProperties, '$.Subject.Job.Bridge') IS NOT NULL OR json_extract(s.JsonProperties, '$.Subject.Model.Type') = 'Hierarchy') THEN 1 ELSE 0 END hideInHierarchy FROM bis.Subject s
```

```
SELECT p.ECInstanceId id, p.Parent.Id parentId FROM bis.InformationPartitionElement p INNER JOIN bis.GeometricModel3d m ON m.ModeledElement.Id = p.ECInstanceId WHERE NOT m.IsPrivate AND EXISTS (SELECT 1 FROM BisCore.GeometricElement3d WHERE Model.Id = m.ECInstanceId)
```

```
SELECT Model.Id modelId, Category.Id categoryId FROM BisCore.GeometricElement3d WHERE Parent IS NULL GROUP BY modelId, categoryId
```

```
SELECT Model.Id modelId, COUNT(*) elementCount FROM BisCore.GeometricElement3d GROUP BY Model.Id
```

3) We have defined our own `FeatureOverrideProvider` which is essentially an instance of `EmphasizeElements` wrapped into another object/subclass of `FeatureOverrideProvider`. This is the solution we came up with to avoid the limitation of having only one instance of `EmphasizeElements` (or any override provider type) targeting a viewport. The idea is that we didn't want to replace/overwrite what is set in the one obtained by calling `EmphasizeElements.getOrCreate(vp)`.

I'd like to mention that some of our tools didn't exit prior to opening the viewport, which led to the emphasizeElements being reapplied (and potentially in both viewport since we have sync logic). It's been taken care of. We get less queryRows calls, but still probably more than we should.

Alex

**Comment by a-gagnon** [comment]
Seems like this could also be related to https://github.com/iTwin/viewer-components-react/issues/1093

**Comment by grigasp** [comment]
> Seems like this could also be related to [iTwin/viewer-components-react#1093](https://github.com/iTwin/viewer-components-react/issues/1093)

Yes, I hope https://github.com/iTwin/viewer-components-react/pull/1098 will have a positive effect on this issue. However, that PR is targeted towards improving performance for the cases when hierarchy is filtered by huge amount of target elements, and determining nodes' visibility in filtered hierarchy is different from unfiltered one, so definitely not all the changes will affect what you're seeing.

I'll try to look into your issue this week.
